Description: Adding two signals that are needed for Application Indicator to keep up-to-date with changed GtkMenu(Item)s.
Author: Cody Russell <bratsche@gnome.org>
Author: Sense Hofstede <sense@ubuntu.com>
Bug-Ubuntu: https://launchpad.net/bugs/608219

Index: gtk+3.0-3.3.4/gtk/gtkcontainer.c
===================================================================
--- gtk+3.0-3.3.4.orig/gtk/gtkcontainer.c	2011-11-19 18:27:36.000000000 +0100
+++ gtk+3.0-3.3.4/gtk/gtkcontainer.c	2011-12-07 18:38:54.157638123 +0100
@@ -248,6 +248,8 @@
   REMOVE,
   CHECK_RESIZE,
   SET_FOCUS_CHILD,
+  CHILD_ADDED,
+  CHILD_REMOVED,
   LAST_SIGNAL
 };
 
@@ -509,6 +511,24 @@
                   NULL, NULL,
                   _gtk_marshal_VOID__OBJECT,
                   G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
+  container_signals[CHILD_ADDED] =
+    g_signal_new (I_("child-added"),
+                  G_OBJECT_CLASS_TYPE (gobject_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
+  container_signals[CHILD_REMOVED] =
+    g_signal_new (I_("child-removed"),
+                  G_OBJECT_CLASS_TYPE (gobject_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
                   GTK_TYPE_WIDGET);
 
   g_type_class_add_private (class, sizeof (GtkContainerPrivate));
Index: gtk+3.0-3.3.4/gtk/gtkmenuitem.c
===================================================================
--- gtk+3.0-3.3.4.orig/gtk/gtkmenuitem.c	2011-12-07 18:38:43.665637615 +0100
+++ gtk+3.0-3.3.4/gtk/gtkmenuitem.c	2011-12-07 18:38:54.165638123 +0100
@@ -86,6 +86,8 @@
   TOGGLE_SIZE_ALLOCATE,
   SELECT,
   DESELECT,
+  SUBMENU_ADDED,
+  SUBMENU_REMOVED,
   LAST_SIGNAL
 };
 
@@ -316,6 +318,25 @@
                   _gtk_marshal_VOID__VOID,
                   G_TYPE_NONE, 0);
 
+  menu_item_signals[SUBMENU_ADDED] =
+    g_signal_new (I_("submenu-added"),
+		  G_OBJECT_CLASS_TYPE (gobject_class),
+		  G_SIGNAL_RUN_FIRST,
+ 		  0,
+		  NULL, NULL,
+		  _gtk_marshal_VOID__OBJECT,
+		  G_TYPE_NONE, 1,
+		  GTK_TYPE_WIDGET);
+ 
+ menu_item_signals[SUBMENU_REMOVED] =
+    g_signal_new (I_("submenu-removed"),
+		  G_OBJECT_CLASS_TYPE (gobject_class),
+		  G_SIGNAL_RUN_FIRST,
+ 		  0,
+		  NULL, NULL,
+		  _gtk_marshal_VOID__VOID,
+		  G_TYPE_NONE, 0);
+
   /**
    * GtkMenuItem:right-justified:
    *
@@ -1321,6 +1342,11 @@
       if (gtk_widget_get_parent (GTK_WIDGET (menu_item)))
         gtk_widget_queue_resize (GTK_WIDGET (menu_item));
 
+      if(submenu == NULL)
+        g_signal_emit_by_name (menu_item, "submenu-removed");
+      else
+        g_signal_emit_by_name (menu_item, "submenu-added", GTK_WIDGET(submenu));
+
       g_object_notify (G_OBJECT (menu_item), "submenu");
     }
 }
Index: gtk+3.0-3.3.4/gtk/gtkmenushell.c
===================================================================
--- gtk+3.0-3.3.4.orig/gtk/gtkmenushell.c	2011-12-07 18:38:43.673637614 +0100
+++ gtk+3.0-3.3.4/gtk/gtkmenushell.c	2011-12-07 18:38:54.165638123 +0100
@@ -643,6 +643,8 @@
                               GTK_WIDGET (menu_shell),
                               child,
                               position);
+
+  g_signal_emit_by_name (menu_shell, "child-added", child);
 }
 
 static void
